dev/tools/deploy-to-kube

```diff
@@ -59,73 +77,158 @@ def walk_object(obj, fn, field_path=""):
+def gather_manifests(manifests_path):
+    """Scans a directory for YAML files and parses them into a structured list.
 
-    manifests_path = os.path.join(repo_root, "k8s")
+    This function walks through the given directory, reads all '.yaml' or '.yml'
+    files, and handles multi-Kubernetes object files by splitting them into individual
+    objects.
 
+    Args:
+        manifests_path (str): The path to the directory containing manifests.
+
+    Returns:
+        list: A flat list of dictionaries. Each dictionary represents a single
+              Kubernetes object and contains the document itself ('doc'), its
+              original 'filename', its 'kind', and the 'root' directory it was found in.
+    """
+    all_docs = []
     if not os.path.isdir(manifests_path):
         print("k8s directory not found")
-        return
-
-    prereq_docs = []
-    other_docs = []
+        return all_docs
 
-    for root, dirs, files in os.walk(manifests_path):
+    for root, _, files in os.walk(manifests_path):
         for filename in files:
             if not (filename.endswith(".yaml") or filename.endswith(".yml")):
                 continue
-
             path = os.path.join(root, filename)
             with open(path, "r") as f:
-                # Use safe_load_all for multi-document YAML files
                 docs = list(yaml.safe_load_all(f))
+                for doc in docs:
+                    if doc:
+                        all_docs.append({
+                            "doc": doc,
+                            "filename": filename,
+                            "kind": doc.get("kind"),
+                        })
+    return all_docs
+
+
+def process_manifests(all_docs, args):
+    """Filters, modifies, and sorts a list of raw manifest objects.
+
+    This function applies the core deployment logic based on the script's
+    arguments. It filters out unwanted documents, replaces image tags, and
+    sorts the documents into three batches for ordered application:
+    prerequisites, standard documents, and extensions documents.
+
+    Args:
+        all_docs (list): The raw list of manifest objects from gather_manifests.
+        args (argparse.Namespace): The parsed command-line arguments.
+
+    Returns:
+        tuple: A tuple containing three lists of documents:
+               (prereq_docs, other_docs, extensions_docs).
+    """
+    prereq_docs = []
+    extensions_docs = []
+    other_docs = []
+
+    image_prefix = utils.get_image_prefix(args)
+    image_tag = utils.get_image_tag()
+
+    for item in all_docs:
+        doc = item["doc"]
+        filename = item["filename"]
+        kind = item["kind"]
+
+        is_extension_file = filename.startswith("extensions")
+        if is_extension_file and not args.extensions:
+            continue
+
+        # When extensions is enabled, we use a the extensions controller manifest instead.
+        if args.extensions and filename == "controller.yaml" and kind == "StatefulSet":
```

Comment:
This is slightly brittle if the file is renamed or if multiple controller files exist. This also assumes file names follow certain convention (e.g. start with extensions). Also, we might change the controller to a Deployment in https://github.com/kubernetes-sigs/agent-sandbox/pull/191. 

Consider identifying the controller by its metadata (e.g., name/labels) rather than the filename, and verifying that an "extension" replacement actually exists before skipping.